### PR TITLE
Port `SegmentationImage` to the new image archetype style

### PR DESCRIPTION
**Pull request by emilk**
### What
* Part of https://github.com/rerun-io/rerun/issues/6386
* Part of https://github.com/rerun-io/rerun/issues/6844
* Very similar to https://github.com/rerun-io/rerun/pull/6915

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6928?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6928?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/6928)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/store/re_types/src/archetypes/segmentation_image_ext.rs [hunk 1]
```diff
@@ -16,62 +17,47 @@ impl SegmentationImage {
...
 
-        match non_empty_dim_inds.len() {
-            2 => {
-                assign_if_none(&mut data.shape[non_empty_dim_inds[0]].name, "height");
-                assign_if_none(&mut data.shape[non_empty_dim_inds[1]].name, "width");
+        if non_empty_dim_inds.len() != 2 {
+            return Err(ImageConstructionError::BadImageShape(tensor_data.shape));
+        }
+
+        let (blob, data_type) = match tensor_data.buffer {
```
Comment (Wumpf) [inline crates/store/re_types/src/archetypes/segmentation_image_ext.rs:30]: maybe now is the time to add a conversion function that is shared between `segmentation_image_ext` and `depth_image_ext`?

File: rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp [hunk 2]
```diff
@@ -1,75 +1,56 @@
...
-    SegmentationImage(Collection<datatypes::TensorDimension> shape, datatypes::TensorBuffer buffer)
-        : SegmentationImage(datatypes::TensorData(std::move(shape), std::move(buffer))) {}
+    /// The length of the data should be `W * H`.
+    template <typename TElement>
+    SegmentationImage(const TElement* pixels, components::Resolution2D resolution_)
+        : SegmentationImage{
+              reinterpret_cast<const uint8_t*>(pixels), resolution_, get_data_type(pixels)} {}
 
-    /// New segmentation image from tensor data.
+    /// Row-major.
```
Comment (Wumpf) [inline rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp:19]: these are pretty sporadic, would be nice to write something nicer imho. Same on DepthImage, but didn't pay enough attention last time.

* what is row major?
* what does "borrows" mean? That's not a common concept in C++ and the implication (crashes at random) needs to be clarified

File: rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp [hunk 3]
```diff
@@ -1,75 +1,56 @@
...
+    /// The length of the data should be `W * H * data_type.size`
+    SegmentationImage(
+        const void* data_, components::Resolution2D resolution_,
+        components::ChannelDataType data_type_
+    )
+        : data{Collection<uint8_t>::borrow(data_, num_bytes(resolution_, data_type_))},
+          resolution{resolution_},
+          data_type{data_type_} {}
+
+    /// New depth image from an `ChannelDataType` and a pointer.
```
Comment (Wumpf) [inline rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp:44]: that's not the variant that takes a pointer, it's the one that takes a (row major) collection!

**Review by Wumpf — APPROVED** [review]

File: rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp (hunk 2 above)
Comment (emilk) [reply rerun_cpp/src/rerun/archetypes/segmentation_image_ext.cpp:19]: I'll add a TODO to https://github.com/rerun-io/rerun/issues/6386

**Pull request merged**